### PR TITLE
Batch multiplier Bug fixes, Added "sum" as loss normalization option

### PR DESCRIPTION
**Pull request by B-Czarnetzki**
Fixed some bugs in training.py regarding batch_multiplier
Added "sum" option for loss normalization

For detailed descrpitions see commits d184815 (bug fixes), fd15837 (Added sum option)

Code for fixing the bug might not be ideal in regards to readability and may be optimizable.

### CODE REVIEW

File: joeynmt/training.py [hunk 1]
```diff
@@ -266,6 +269,15 @@ def train_and_validate(self, train_data: Dataset, valid_data: Dataset) \
                                     batch_size=self.batch_size,
                                     batch_type=self.batch_type,
                                     train=True, shuffle=self.shuffle)
+
+        # For last batch in epoch batch_multiplier needs to be adjusted to fit the number of leftover training examples
+        leftover_batch_size = len(
+            train_data) % (self.batch_multiplier * self.batch_size)
+
+        # Problem occurs if count can't be counted down to 0
+        batch_mutliplier_problem = self.batch_size * \
```
Comment (juliakreutzer) [inline joeynmt/training.py]: Typo -> multiplier

File: joeynmt/training.py [hunk 2]
```diff
@@ -291,14 +304,29 @@ def train_and_validate(self, train_data: Dataset, valid_data: Dataset) \
...
                 # print(count, update, self.steps)
-                batch_loss = self._train_batch(batch, update=update)
-                self.tb_writer.add_scalar("train/train_batch_loss", batch_loss,
-                                          self.steps)
+                batch_loss = self._train_batch(
+                    batch, update=update, count=count)
+
+                # Only save finaly computed batch_loss of full batch
+                if update:
+                    self.tb_writer.add_scalar("train/train_batch_loss", batch_loss,
```
Comment (juliakreutzer) [inline joeynmt/training.py]: Good catch! Thanks

File: joeynmt/training.py [hunk 3]
```diff
@@ -410,21 +438,21 @@ def train_and_validate(self, train_data: Dataset, valid_data: Dataset) \
...
         self.logger.info('Best validation result (greedy) at step '
                          '%8d: %6.2f %s.', self.best_ckpt_iteration,
                          self.best_ckpt_score,
                          self.early_stopping_metric)
 
         self.tb_writer.close()  # close Tensorboard writer
 
-    def _train_batch(self, batch: Batch, update: bool = True) -> Tensor:
+    def _train_batch(self, batch: Batch, update: bool = True, count: int = 1) -> Tensor:
         """
```
Comment (juliakreutzer) [inline joeynmt/training.py]: Could you please add the new parameter to the docstrings?

File: joeynmt/training.py [hunk 4]
```diff
@@ -440,28 +468,39 @@ def _train_batch(self, batch: Batch, update: bool = True) -> Tensor:
             normalizer = batch.nseqs
         elif self.normalization == "tokens":
             normalizer = batch.ntokens
+        elif self.normalization == "sum":
```
Comment (juliakreutzer) [inline joeynmt/training.py]: I'd rather call it "none", since summing is not a normalization technique.

File: joeynmt/training.py [hunk 5]
```diff
@@ -440,28 +468,39 @@ def _train_batch(self, batch: Batch, update: bool = True) -> Tensor:
...
 
             # increment step counter
             self.steps += 1
 
+        else:
+            if count == self.current_batch_multiplier - 1:
+                self.norm_batch_loss_accumulated = norm_batch_loss
+            else:
+                # accumalte loss of current batch_size * batch_multiplier loss
+                self.norm_batch_loss_accumulated += norm_batch_loss
```
Comment (juliakreutzer) [inline joeynmt/training.py]: Type "accumulate"

File: joeynmt/training.py [hunk 6]
```diff
@@ -67,9 +69,9 @@ def __init__(self, model: Model, config: dict) -> None:
         self.loss = XentLoss(pad_index=self.pad_index,
                              smoothing=self.label_smoothing)
         self.normalization = train_config.get("normalization", "batch")
-        if self.normalization not in ["batch", "tokens"]:
-            raise ConfigurationError("Invalid normalization. "
-                                     "Valid options: 'batch', 'tokens'.")
+        if self.normalization not in ["batch", "tokens", "sum"]:
+            raise ConfigurationError("Invalid normalization option."
+                                     "Valid options: 'batch', 'tokens', 'sum'.")
```
Comment (juliakreutzer) [inline joeynmt/training.py]: Let's call it "none", see below.

File: joeynmt/training.py [hunk 7]
```diff
@@ -291,14 +304,29 @@ def train_and_validate(self, train_data: Dataset, valid_data: Dataset) \
                 # see https://medium.com/@davidlmorton/
                 # increasing-mini-batch-size-without-increasing-
                 # memory-6794e10db672
+
+                # Set current_batch_mutliplier to fit number of leftover examples for last batch in epoch
+                if self.batch_multiplier > 1 and batch_mutliplier_problem and i == len(train_iter) - math.ceil(leftover_batch_size / self.batch_size):
```
Comment (juliakreutzer) [inline joeynmt/training.py]: I think we can simplify this. We only need to check whether the batch_multiplier is used and whether it's the last batch. Then we retrieve its size and scale the batch_multiplier accordingly. If there is no problem, the scaling should return 1.

Comment (B-Czarnetzki) [reply joeynmt/training.py]: I deleted the "batch_multiplier_problem' computation, since it was redundant. 368b6b0
I don't think we can simplify it further since we need the number of leftover examples (leftover_batch_size) anyway to get the last actual batch (batch_size * batch_mutliplier).
Scaling self.batch_multiplier itself, instead of computing self.current_batch_mutliplier, isn't sensible since it would than be incorrect for the next epoch.

Comment (juliakreutzer) [reply joeynmt/training.py]: Yes, you're right. Thanks!

**Comment by juliakreutzer** [comment]
Awesome, great work @B-Czarnetzki! Thanks a lot :koala:
